src/vmm_asn.ml

```diff
@@ -683,18 +762,66 @@ let unikernel_info =
+                                   (required ~label:"unikernel-device" utf8_string)
+                                   (optional ~label:"host-device" utf8_string)
                                    (optional ~label:"mac" mac_addr)))))
          @ (optional ~label:"arguments"(my_explicit 2 (sequence_of utf8_string)))
         -@ (optional ~label:"started" (my_explicit 3 generalized_time)))
 
+let unikernel_info =
+  let open Unikernel in
+  let f (typ, (fail_behaviour, (cpuid, (memory, (digest, (blocks, (bridges, (argv, started)))))))) =
+    let bridges = match bridges with None -> [] | Some xs ->
+      List.map (fun (unikernel_device, host_device, mac) ->
+          { unikernel_device ; host_device ; mac })
+        xs
+    and block_devices = match blocks with None -> [] | Some xs ->
+      List.map (fun (unikernel_device, host_device, sector_size, size) ->
+          { unikernel_device ; host_device ; sector_size ; size })
+        xs
+    and started = Option.value ~default:Ptime.epoch started
+    in
+    { typ ; fail_behaviour ; cpuid ; memory ; block_devices ; bridges ; argv ; digest ; started }
+  and g (unikernel : info) =
+    let bridges = match unikernel.bridges with
+      | [] -> None
+      | xs -> Some (List.map (fun { unikernel_device ; host_device ; mac } ->
+          unikernel_device, host_device, mac) xs)
+    and blocks = match unikernel.block_devices with
+      | [] -> None
+      | xs -> Some (List.map (fun { unikernel_device ; host_device ; sector_size ; size } ->
+          unikernel_device, host_device, sector_size, size) xs)
+    in
+    (unikernel.typ, (unikernel.fail_behaviour, (unikernel.cpuid, (unikernel.memory, (unikernel.digest, (blocks, (bridges, (unikernel.argv, Some unikernel.started))))))))
+  in
+  Asn.S.(map f g @@ sequence @@
+           (required ~label:"typ" typ)
+         @ (required ~label:"fail-behaviour" fail_behaviour)
+         @ (required ~label:"cpuid" int)
+         @ (required ~label:"memory" int)
+         @ (required ~label:"digest" octet_string)
+         @ (optional ~label:"blocks"
+              (my_explicit 0 (set_of
+                                (sequence4
+                                   (required ~label:"unikernel-device" utf8_string)
+                                   (required ~label:"host-device" utf8_string)
+                                   (required ~label:"block-sector-size" int)
+                                   (required ~label:"block-size" int)))))
+         @ (optional ~label:"bridges"
+              (my_explicit 1 (set_of
+                                (sequence3
+                                   (required ~label:"unikernel-device" utf8_string)
+                                   (required ~label:"host-device" utf8_string)
+                                   (required ~label:"mac" mac_addr)))))
+         @ (optional ~label:"arguments"(my_explicit 2 (sequence_of utf8_string)))
+        -@ (optional ~label:"started" (my_explicit 3 generalized_time)))
```

Comment:
Why are these optional?

Reply:
This is not new.

Reply:
I guess backwards compatibility is the answer. Making things optional means that the old format (without arguments/started) is still a valid thing.